File: backend/app/services/feature_importance.py
"""Per-disease feature importance derived from trained models with safe fallbacks."""
from __future__ import annotations

from typing import Any

import numpy as np

from app.services.features import FEATURES
from app.services.registry import registry

# Canonical fallbacks if no model is loaded (educational priors, sum ≈ 1.0).
FALLBACK_IMPORTANCE: dict[str, dict[str, float]] = {
    "diabetes": {"Glucose": 0.35, "BMI": 0.22, "Age": 0.18, "Insulin": 0.15, "BloodPressure": 0.10},
    "heart": {"Cholesterol": 0.28, "RestingBP": 0.22, "MaxHR": 0.18, "Age": 0.17, "STDepression": 0.15},
    "kidney": {"SerumCreatinine": 0.32, "BloodUrea": 0.24, "Hemoglobin": 0.20, "Albumin": 0.14, "BloodPressure": 0.10},
    "liver": {"TotalBilirubin": 0.30, "DirectBilirubin": 0.22, "SGPT": 0.20, "SGOT": 0.16, "AlkalinePhosphotase": 0.12},
}


def compute_importance(disease: str, top_k: int = 8) -> list[dict[str, Any]]:
    feature_names = FEATURES.get(disease, [])
    model = registry.best_for(disease)
    importances: np.ndarray | None = None
    if model is not None:
        est = model.estimator
        if hasattr(est, "feature_importances_"):
            importances = np.asarray(est.feature_importances_, dtype=float)
        elif hasattr(est, "coef_"):
            importances = np.abs(np.ravel(est.coef_)).astype(float)
    if importances is not None and importances.sum() > 0 and len(importances) == len(feature_names):
        total = float(importances.sum())
        pairs = [(n, float(v) / total) for n, v in zip(feature_names, importances)]
    else:
        pairs = [(k, v) for k, v in FALLBACK_IMPORTANCE.get(disease, {}).items()]
    pairs.sort(key=lambda p: p[1], reverse=True)
    return [{"name": n, "importance": round(v, 4)} for n, v in pairs[:top_k]]

File: src/lib/feature-importance.ts
/**
 * Per-disease feature importance. Tries the FastAPI backend; falls back to
 * canonical educational priors so the UI always has data to render.
 */
import type { DiseaseKey } from "./predict-api";
import { PREDICT_API_BASE } from "./predict-api";

export type FeatureImportance = { name: string; importance: number };

export const FALLBACK_IMPORTANCE: Record<DiseaseKey, FeatureImportance[]> = {
  diabetes: [
    { name: "Glucose", importance: 0.35 },
    { name: "BMI", importance: 0.22 },
    { name: "Age", importance: 0.18 },
    { name: "Insulin", importance: 0.15 },
    { name: "BloodPressure", importance: 0.10 },
  ],
  heart: [
    { name: "Cholesterol", importance: 0.28 },
    { name: "RestingBP", importance: 0.22 },
    { name: "MaxHR", importance: 0.18 },
    { name: "Age", importance: 0.17 },
    { name: "STDepression", importance: 0.15 },
  ],
  kidney: [
    { name: "SerumCreatinine", importance: 0.32 },
    { name: "BloodUrea", importance: 0.24 },
    { name: "Hemoglobin", importance: 0.20 },
    { name: "Albumin", importance: 0.14 },
    { name: "BloodPressure", importance: 0.10 },
  ],
  liver: [
    { name: "TotalBilirubin", importance: 0.30 },
    { name: "DirectBilirubin", importance: 0.22 },
    { name: "SGPT", importance: 0.20 },
    { name: "SGOT", importance: 0.16 },
    { name: "AlkalinePhosphotase", importance: 0.12 },
  ],
};

export async function fetchFeatureImportance(disease: DiseaseKey): Promise<FeatureImportance[]> {
  try {
    const res = await fetch(`${PREDICT_API_BASE}/feature-importance/${disease}`);
    if (!res.ok) throw new Error(String(res.status));
    const j = (await res.json()) as { features: FeatureImportance[] };
    if (Array.isArray(j.features) && j.features.length) return j.features;
  } catch {
    /* fall through */
  }
  return FALLBACK_IMPORTANCE[disease];
}
